reshuffle_test.go

```diff
@@ -113,3 +114,196 @@ func TestRepartitionType(t *testing.T) {
+			func(shard int, _ struct{}, _ error, xs []int) error {
+				rowsPerShard[shard] += len(xs)
+				return nil
+			},
+		)
+	}
+
+	const numShards = 2
+	slice := bigslice.Const(numShards, []int{1, 2, 3, 4, 5, 6})
+
+	slice0 := countRowsPerShard(numShards, slice)
+	fmt.Println("# default partitioning")
+	fmt.Println("## slice contents")
+	slicetest.Print(slice0)
+	fmt.Println("## row count per shard")
+	for shard, count := range rowsPerShard {
+		fmt.Printf("shard:%d count:%d\n", shard, count)
+	}
+
+	slice1 := bigslice.Repartition(slice, func(nshard, x int) int {
+		// We know our slice keys are sequential integers, so we partition
+		// perfectly with mod.
+		return x % nshard
+	})
+	slice1 = countRowsPerShard(numShards, slice1)
+	fmt.Println("# repartitioned")
+	// Note that the slice contents are unchanged.
+	fmt.Println("## slice contents")
+	slicetest.Print(slice1)
+	// Note that the partitioning has changed.
+	fmt.Println("## row count per shard")
+	for shard, count := range rowsPerShard {
+		fmt.Printf("shard:%d count:%d\n", shard, count)
+	}
+	// Output:
+	// # default partitioning
+	// ## slice contents
+	// 1
+	// 2
+	// 3
+	// 4
+	// 5
+	// 6
+	// ## row count per shard
+	// shard:0 count:4
+	// shard:1 count:2
+	// # repartitioned
+	// ## slice contents
+	// 1
+	// 2
+	// 3
+	// 4
+	// 5
+	// 6
+	// ## row count per shard
+	// shard:0 count:3
+	// shard:1 count:3
+}
+
+func ExampleReshard() {
+	// countRowsPerShard is a utility that counts the number of rows per shard
+	// and stores it in rowsPerShard.
+	var rowsPerShard []int
+	countRowsPerShard := func(numShards int, slice bigslice.Slice) bigslice.Slice {
+		rowsPerShard = make([]int, numShards)
+		return bigslice.WriterFunc(slice,
+			func(shard int, _ struct{}, _ error, xs []int) error {
+				rowsPerShard[shard] += len(xs)
+				return nil
+			},
+		)
+	}
+
+	const beforeNumShards = 2
+	slice := bigslice.Const(beforeNumShards, []int{1, 2, 3, 4, 5, 6})
+
+	before := countRowsPerShard(beforeNumShards, slice)
+	fmt.Println("# before")
+	fmt.Println("## slice contents")
+	slicetest.Print(before)
+	fmt.Println("## row count per shard")
+	for shard, count := range rowsPerShard {
+		fmt.Printf("shard:%d count:%d\n", shard, count)
+	}
+
+	// Reshard to 4 shards.
+	const afterNumShards = 4
+	after := bigslice.Reshard(slice, afterNumShards)
+	after = countRowsPerShard(afterNumShards, after)
+	fmt.Println("# after")
+	fmt.Println("## slice contents")
+	slicetest.Print(after)
+	fmt.Println("## row count per shard")
+	for shard, count := range rowsPerShard {
+		fmt.Printf("shard:%d count:%d\n", shard, count)
+	}
+	// Output:
+	// # before
+	// ## slice contents
+	// 1
+	// 2
+	// 3
+	// 4
+	// 5
+	// 6
+	// ## row count per shard
+	// shard:0 count:4
+	// shard:1 count:2
+	// # after
+	// ## slice contents
+	// 1
+	// 2
+	// 3
+	// 4
+	// 5
+	// 6
+	// ## row count per shard
+	// shard:0 count:2
+	// shard:1 count:1
+	// shard:2 count:1
+	// shard:3 count:2
+}
+
+func ExampleReshuffle() {
```

Comment:
Some of these examples are pretty detailed in setup. I think that's fine; readers will come here after they're already invested. But to make it a little easier, how about a quick (1 sentence) summary of the plan in a comment at the top, just so they have a preview of that they'll be looking for? For the ones that are longer than ~1 screen.

"ExampleReshuffle counts rows per shard before and after a reshuffle, showing same-keyed rows all go to the same shard."

Reply:
I gave it a whirl.